logstash-core/lib/logstash/config/source_loader.rb

```diff
@@ -0,0 +1,68 @@
+# encoding: utf-8
+require "logstash/config/source/local"
+require "logstash/errors"
+require "thread"
+require "set"
+
+module LogStash module Config
+  class SourceLoader
+    class Loader
+      def initialize(loaders)
+        @loaders = loaders
+      end
+
+      def fetch
+        @loaders.collect(&:pipeline_configs).flatten
+      end
+    end
+
+    include LogStash::Util::Loggable
+
+    def initialize
+      @source_loaders_mutex = Mutex.new
+      @source_loaders = Set.new([LogStash::Config::Source::Local])
+    end
+
+    # This return a ConfigLoader object that will
+    # abstract the call to the different sources and will return multiples pipeline
+    def create(settings)
+      loaders = []
+
+      source_loaders do |config_source|
+        if config_source.match?(settings)
+          loaders << config_source.new(settings)
+        end
+      end
+
+      if loaders.empty?
+
+        # This shouldn't happen with the settings object or with any external plugins.
+        # but lets add a guard so we fail fast.
+        raise LogStash::InvalidSourceLoaderSettingError, "Can't find an appropriate config loader with current settings"
+      else
+        Loader.new(loaders)
```

Comment:
In order to dynamically add and remove sources to the source loader, we could remove this class and use instances of SourceLoader directly